#### Task 1 Problem Statement: Build machine learning models for prediction

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import metrics as m


In [2]:
df=pd.read_csv('House_Prediction.csv')
df

Unnamed: 0   price   lotsize  bedrooms  bathrms  stories  garagepl  \
0             0   42000  8.674197       3.0        1        2         1   
1             1   38500  8.294050       2.0        1        1         0   
2             2   49500  8.026170       3.0        1        1         0   
3             3   60500  8.802372       3.0        1        2         0   
4             4   61000  8.757784       3.0        1        1         0   
..          ...     ...       ...       ...      ...      ...       ...   
546         546  107500  8.699515       3.0        2        4         1   
547         547  108000  8.699515       3.0        2        3         0   
548         548  113750  8.699515       3.0        1        4         2   
549         549  120000  8.853665       3.0        1        4         2   
550         550   70000  9.464983       3.0        1        1         2   

     driveway_yes  recroom_yes  fullbase_yes  gashw_yes  airco_yes  \
0               1            0             1          0          0   
1               1            0             0          0          0   
2               1            0             0          0          0   
3               1            1             0          0          0   
4               1            0             0          0          0   
..            ...          ...           ...        ...        ...   
546             1            0             0          0          1   
547             1            0             0          0          1   
548             1            1             0          0          1   
549             1            0             0          0          1   
550             1            0             0          0          0   

     prefarea_yes  
0               0  
1               0  
2               0  
3               0  
4               0  
..            ...  
546             0  
547             0  
548             0  
549             0  
550             0  

[551 rows x 13 columns]

This dataset shows features that contribute to the pricing of a house. The dependent variable is the price

In [3]:
df.head()

Unnamed: 0  price   lotsize  bedrooms  bathrms  stories  garagepl  \
0           0  42000  8.674197       3.0        1        2         1   
1           1  38500  8.294050       2.0        1        1         0   
2           2  49500  8.026170       3.0        1        1         0   
3           3  60500  8.802372       3.0        1        2         0   
4           4  61000  8.757784       3.0        1        1         0   

   driveway_yes  recroom_yes  fullbase_yes  gashw_yes  airco_yes  prefarea_yes  
0             1            0             1          0          0             0  
1             1            0             0          0          0             0  
2             1            0             0          0          0             0  
3             1            1             0          0          0             0  
4             1            0             0          0          0             0

In [4]:
#statistical values of the data
df.describe()

Unnamed: 0          price     lotsize    bedrooms     bathrms  \
count  551.000000     551.000000  551.000000  551.000000  551.000000   
mean   275.000000   68445.811252    8.470413    2.967332    1.286751   
std    159.204271   26848.486040    0.399086    0.732880    0.502165   
min      0.000000   25000.000000    7.408531    1.000000    1.000000   
25%    137.500000   49500.000000    8.188689    3.000000    1.000000   
50%    275.000000   62500.000000    8.433812    3.000000    1.000000   
75%    412.500000   82950.000000    8.757784    3.000000    2.000000   
max    550.000000  190000.000000    9.692767    6.000000    4.000000   

          stories    garagepl  driveway_yes  recroom_yes  fullbase_yes  \
count  551.000000  551.000000    551.000000   551.000000    551.000000   
mean     1.820327    0.698730      0.860254     0.177858      0.346642   
std      0.881334    0.863386      0.347038     0.382741      0.476333   
min      1.000000    0.000000      0.000000     0.000000      0.000000   
25%      1.000000    0.000000      1.000000     0.000000      0.000000   
50%      2.000000    0.000000      1.000000     0.000000      0.000000   
75%      2.000000    1.000000      1.000000     0.000000      1.000000   
max      4.000000    3.000000      1.000000     1.000000      1.000000   

        gashw_yes   airco_yes  prefarea_yes  
count  551.000000  551.000000    551.000000  
mean     0.045372    0.321234      0.232305  
std      0.208308    0.467375      0.422686  
min      0.000000    0.000000      0.000000  
25%      0.000000    0.000000      0.000000  
50%      0.000000    0.000000      0.000000  
75%      0.000000    1.000000      0.000000  
max      1.000000    1.000000      1.000000

This shows the statistical values of the dataset with the mean price of a house being 68445.811252 and the max price being 190000

In [5]:
#information of data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 551 entries, 0 to 550
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    551 non-null    int64  
 1   price         551 non-null    int64  
 2   lotsize       551 non-null    float64
 3   bedrooms      551 non-null    float64
 4   bathrms       551 non-null    int64  
 5   stories       551 non-null    int64  
 6   garagepl      551 non-null    int64  
 7   driveway_yes  551 non-null    int64  
 8   recroom_yes   551 non-null    int64  
 9   fullbase_yes  551 non-null    int64  
 10  gashw_yes     551 non-null    int64  
 11  airco_yes     551 non-null    int64  
 12  prefarea_yes  551 non-null    int64  
dtypes: float64(2), int64(11)
memory usage: 56.1 KB


The information being displayed in the dataset is that there are a total of 551 entries starting from index 0-550. Therefore there are 551 rows and 13 columns.There are non-null values and all the data is numerical.

In [6]:
#correlation of the data against the dependant variable price
df.corr() 

Unnamed: 0     price   lotsize  bedrooms   bathrms   stories  \
Unnamed: 0      1.000000  0.387924  0.386297  0.109361  0.109914  0.248420   
price           0.387924  1.000000  0.560017  0.363247  0.513014  0.435332   
lotsize         0.386297  0.560017  1.000000  0.151814  0.198791  0.112181   
bedrooms        0.109361  0.363247  0.151814  1.000000  0.371325  0.399058   
bathrms         0.109914  0.513014  0.198791  0.371325  1.000000  0.322034   
stories         0.248420  0.435332  0.112181  0.399058  0.322034  1.000000   
garagepl        0.135595  0.385734  0.365816  0.136709  0.170263  0.052983   
driveway_yes    0.315854  0.298859  0.332750 -0.010833  0.042566  0.125817   
recroom_yes     0.095841  0.253611  0.176168  0.079088  0.122017  0.046397   
fullbase_yes   -0.013450  0.175110  0.035777  0.094997  0.100567 -0.180526   
gashw_yes      -0.036458  0.089257 -0.015737  0.045456  0.066592  0.014774   
airco_yes       0.169776  0.462458  0.262216  0.158087  0.187823  0.312520   
prefarea_yes    0.503575  0.318696  0.212355  0.077366  0.062495  0.034156   

              garagepl  driveway_yes  recroom_yes  fullbase_yes  gashw_yes  \
Unnamed: 0    0.135595      0.315854     0.095841     -0.013450  -0.036458   
price         0.385734      0.298859     0.253611      0.175110   0.089257   
lotsize       0.365816      0.332750     0.176168      0.035777  -0.015737   
bedrooms      0.136709     -0.010833     0.079088      0.094997   0.045456   
bathrms       0.170263      0.042566     0.122017      0.100567   0.066592   
stories       0.052983      0.125817     0.046397     -0.180526   0.014774   
garagepl      1.000000      0.205116     0.041400      0.046609   0.066032   
driveway_yes  0.205116      1.000000     0.091646      0.040602  -0.012735   
recroom_yes   0.041400      0.091646     1.000000      0.369287  -0.010181   
fullbase_yes  0.046609      0.040602     0.369287      1.000000   0.006119   
gashw_yes     0.066032     -0.012735    -0.010181      0.006119   1.000000   
airco_yes     0.159165      0.109127     0.137408      0.037930  -0.131303   
prefarea_yes  0.087499      0.196923     0.159972      0.231448  -0.057977   

              airco_yes  prefarea_yes  
Unnamed: 0     0.169776      0.503575  
price          0.462458      0.318696  
lotsize        0.262216      0.212355  
bedrooms       0.158087      0.077366  
bathrms        0.187823      0.062495  
stories        0.312520      0.034156  
garagepl       0.159165      0.087499  
driveway_yes   0.109127      0.196923  
recroom_yes    0.137408      0.159972  
fullbase_yes   0.037930      0.231448  
gashw_yes     -0.131303     -0.057977  
airco_yes      1.000000      0.109357  
prefarea_yes   0.109357      1.000000

Lotsize is the most correlated with price, that is, lotsize greatly determines the housing price. The larger the lotsize, the higher the price. Number of bathrooms comes in second, then the availability of aircon,and, number of the availability of gashw has the least correlation with price.

In [7]:
print(df.columns)
print(len(df.columns))

Index(['Unnamed: 0', 'price', 'lotsize', 'bedrooms', 'bathrms', 'stories',
       'garagepl', 'driveway_yes', 'recroom_yes', 'fullbase_yes', 'gashw_yes',
       'airco_yes', 'prefarea_yes'],
      dtype='object')
13


In [8]:
#Dropping some columns that have little correlation so that we can split our X variables as (lotsize,bedrooms,bathrooms and stories) and Y variable as Price
#Dropping these columns reduces the risk of overfitting
d = df.drop(df.columns[[0, 8, 9, 10]], axis=1)
d

price   lotsize  bedrooms  bathrms  stories  garagepl  driveway_yes  \
0     42000  8.674197       3.0        1        2         1             1   
1     38500  8.294050       2.0        1        1         0             1   
2     49500  8.026170       3.0        1        1         0             1   
3     60500  8.802372       3.0        1        2         0             1   
4     61000  8.757784       3.0        1        1         0             1   
..      ...       ...       ...      ...      ...       ...           ...   
546  107500  8.699515       3.0        2        4         1             1   
547  108000  8.699515       3.0        2        3         0             1   
548  113750  8.699515       3.0        1        4         2             1   
549  120000  8.853665       3.0        1        4         2             1   
550   70000  9.464983       3.0        1        1         2             1   

     airco_yes  prefarea_yes  
0            0             0  
1            0             0  
2            0             0  
3            0             0  
4            0             0  
..         ...           ...  
546          1             0  
547          1             0  
548          1             0  
549          1             0  
550          0             0  

[551 rows x 9 columns]

In [9]:
#Defining the features and target variable
x=d.iloc[:,1:9].values
y=d.iloc[:,0].values

In [10]:
print(x.shape)
print(y.shape)

(551, 8)
(551,)


In [11]:
#Spliting data into Training and Testing
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0 )

In [12]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(440, 8)
(111, 8)
(440,)
(111,)


In [13]:
print('x_train')
print( x_train)

x_train
[[8.28601747 3.         1.         ... 0.         0.         0.        ]
 [8.69951475 3.         1.         ... 1.         1.         0.        ]
 [8.98469369 3.         1.         ... 1.         0.         1.        ]
 ...
 [8.61250337 3.         2.         ... 1.         1.         0.        ]
 [8.79482493 3.         1.         ... 1.         1.         1.        ]
 [8.15478757 3.         1.         ... 0.         0.         0.        ]]


In [14]:
print('y_train')
print( y_train)

y_train
[ 63000 113750  73500  78900  48000  62600  86000  42000  42000  50000
  97000  90000  50000  92000  47000  64500  74500  74500  60000  89900
 117000  76000 105000  80000  52000  31900  35000 114000  69000 112500
  75000  34000  85000  90000  79000  59900  57000  38000  45000  52900
  54800  57000  93000  49000  93000  45000  25245  57250  48000  66000
  83800  75000  57500 106000  70800  46200  82000 120000  99000 104900
  58550  66000  50000  56000  70000  48000  70000  30000  70000  64900
  64000 107500 107500  48900  33000  25000  32000  43000  60000  62000
  40000  61000  89000  45000  68500  48500  71000  68000  35000  43500
  46000 163000  38000  61000 140000  50000  72000  86900  45000  67000
  66000  40000 120000  53000  90000 120000  37000  37900  47000  40500
  48500  53000  62500  37000  52500  51500  45000 103000  46000  54000
  56000 113000 118500  61500  35000  59000 175000  27000 105000  84000
  70000  63900  91500  26000 132000  83900  64000  62900 112000  5700

In [15]:
#Model Training/Building
from sklearn.linear_model import LinearRegression
linreg=LinearRegression()
linreg.fit(x_train,y_train)# Model buiding | model training

#X_TRAIN = 440 RECORDS, 8 COLUMNS- lotsize, bedrooms, bathrms, stories, garagepl, driveway_yes, airco_yes, prefarea_yes
#Y_TRAIN= 440 RECORDS, 1 COLUMN - price

LinearRegression()

###  Interpreting Model Coefficients

In [16]:
print(linreg.intercept_) #b

-176180.5950513676


In [17]:
print(linreg.coef_) #a1- coeff of lotsize,a2- coeff of bedrooms,a3- coeff of bathrms,a4- coeff of stories, a5- coeff of garagepl, a6- coeff of driveway_yes, a7- coeff of airco_yes, a8- coeff of prefarea_ye

[22759.23836637  3020.99407539 14867.64622586  5413.27276603
  3883.27100792  4909.45167174 12769.75324162 12995.95481011]


In [18]:
print("x_test")
print(x_test)


x_test
[[8.77647579 3.         1.         2.         1.         0.
  1.         0.        ]
 [8.92025541 3.         2.         3.         1.         1.
  0.         1.        ]
 [9.01796848 3.         1.         1.         2.         1.
  0.         0.        ]
 [7.79564654 3.         1.         1.         0.         0.
  0.         0.        ]
 [8.23217424 3.         1.         1.         2.         1.
  0.         0.        ]
 [8.55062797 3.         1.         4.         0.         1.
  1.         0.        ]
 [8.52118521 3.         1.         4.         0.         1.
  1.         1.        ]
 [8.70946508 3.         1.         1.         0.         1.
  0.         0.        ]
 [8.30399997 3.         1.         2.         1.         1.
  0.         0.        ]
 [8.98719682 3.         1.         1.         2.         1.
  1.         1.        ]
 [8.87626526 3.         1.         1.         2.         1.
  0.         1.        ]
 [8.5754621  4.         2.         1.         0.         1

In [19]:
#y=a1x1 +a2x2 +a3x3 +a4x4 +a5x5 +a6x6 +a7x7 +a8x8 +b
print("y_test: (price = actual values)")
print(y_test)

y_test: (price = actual values)
[ 57000 114900  53900  38000  65000  67000  96000  47000  64000  86000
  84000  75000  60000  94500 113750  55000  49000  87000 124000  56000
  99000  54000  75000  83000  32500  52000 120000  56000  50000  95500
  60000  42500  57500  60000  75000  62000  37200  38000  37900 145000
 108000  60000  40750  67000  67000  42000  95000  44000  79500  25000
  85000  44700  87250  58500 106000  85000  72500  70000  69000  95000
  73000  80000  73000  44000  87000  41000 125000  49000  65900  58500
 126500  42000 132000  51000  36000  80000  55000  34000  60000  83900
  50000  51000  75000  53900  60000  47500  60000  59500  88000  75000
  61500  70000  70000 102000  65000  65000  63000  79000  38500  86900
  27000  49900  53000  80000  51900  30000  54000  45000 120000  78000
 106500]


Making Predictions With our Model

In [20]:
y_pred=linreg.predict(x_test)
#x_test = lotsize, bedrooms, bathrms, stories, garagepl, driveway_yes, airco_yes, prefarea_yes ,
#y_test(actual values) = price

In [21]:
print(y_pred)

[ 74975.50768865 103664.39469115  71081.39405571  30586.28390708
  53197.31557361  81688.10947265  94013.96960772  56293.54961639
  56362.01634566  96146.76259291  80852.2915188   96898.0922166
  70281.96337431 113320.14206139  92843.20122022  42948.13085138
  55531.31474964  85473.12569241  95477.83730136  83377.26973451
 101923.61023145  41420.09145435 102670.4623352   74665.93436954
  35716.23318752  42121.56282297  96351.55328674  45704.85852537
  45554.84831795  99318.70002425  70382.19479491  56166.6370989
  72677.42096889  56678.89267281  69476.28916823  55028.33702308
  31667.91164861  44531.78896669  51402.25775751 146735.84049218
  94531.03266421  60362.25313882  66657.76106128  94107.48516704
  57809.77474302  48828.17763195 109066.84678248  33318.90448909
  90974.71462989  34688.90572401  91778.59254129  54066.07536814
  59773.97105485  64245.62162315 108949.54917067  91493.34906838
  83137.3952971   81255.09476157  98067.79871271  94531.03266421
  86083.38225025  75596.641

In [22]:
print(y_test)

[ 57000 114900  53900  38000  65000  67000  96000  47000  64000  86000
  84000  75000  60000  94500 113750  55000  49000  87000 124000  56000
  99000  54000  75000  83000  32500  52000 120000  56000  50000  95500
  60000  42500  57500  60000  75000  62000  37200  38000  37900 145000
 108000  60000  40750  67000  67000  42000  95000  44000  79500  25000
  85000  44700  87250  58500 106000  85000  72500  70000  69000  95000
  73000  80000  73000  44000  87000  41000 125000  49000  65900  58500
 126500  42000 132000  51000  36000  80000  55000  34000  60000  83900
  50000  51000  75000  53900  60000  47500  60000  59500  88000  75000
  61500  70000  70000 102000  65000  65000  63000  79000  38500  86900
  27000  49900  53000  80000  51900  30000  54000  45000 120000  78000
 106500]


In [23]:
d=pd.DataFrame({'Actual':y_test,'Predicted': y_pred})
print(d)

     Actual      Predicted
0     57000   74975.507689
1    114900  103664.394691
2     53900   71081.394056
3     38000   30586.283907
4     65000   53197.315574
..      ...            ...
106   54000   49549.238915
107   45000   27189.092049
108  120000   96561.270067
109   78000   88880.603551
110  106500   72649.163796

[111 rows x 2 columns]


There is a large difference between most of the predicted values and the actual values. 

### Model Evaluation Metrics

In [24]:
m.mean_absolute_error(y_test,y_pred)

12392.031710919453

In [25]:
m.mean_squared_error(y_test,y_pred)

258312717.582496

In [26]:
np.sqrt(m.mean_squared_error(y_test,y_pred ))

16072.109929393091

In [27]:
y_train_pred= linreg.predict(x_train) # training dataset(x_train and y_train)
print("RMSE:",np.sqrt(m.mean_squared_error(y_train,y_train_pred)))# Error on training dataset
print("R-SQUARED:",m.r2_score(y_train, y_train_pred)) # accuracy on training dataset

y_test_pred= linreg.predict(x_test)
print("RMSE:",np.sqrt(m.mean_squared_error(y_test,y_test_pred))) 
print("R-SQUARED:",m.r2_score(y_test, y_test_pred))# accuracy on testing dataset

RMSE: 15752.862002793583
R-SQUARED: 0.6661175038146496
RMSE: 16072.109929393091
R-SQUARED: 0.587000190498637


Variance=15752.862002793583(bias of training) - 16072.109929393091(Bias of testing)
We should aim for low bias and low variance
R-squared coefficient of determination is the value of checking the error in our model. This is higher in the training data than it its in the unseen data.

In [28]:
print(linreg.intercept_)

-176180.5950513676


In [29]:
print(linreg.coef_)

[22759.23836637  3020.99407539 14867.64622586  5413.27276603
  3883.27100792  4909.45167174 12769.75324162 12995.95481011]


In [30]:
m.r2_score(y_test,y_pred)

0.587000190498637

The r2 score is 0.587 (58/59%)meaning the value of checking error is very average.

10% of the price standard deviation (26848.486040) is2684.8486040 which is less than both the MAE and RMSE (RMSE: 16072.109929393091). Therefore, unfortunately, this is not a much reliable model as it is on the average side. However, it is usable. 

### Task 2 GridSearchCV_ Regression

In [31]:
models=[]   # model[] is empty
scores=[]   #scores[] is empty

In [32]:
# Import relavent libraries
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV

# Define the parameters for grid search
parameters = {'fit_intercept': [True, False]}

# Initialize the LinearRegression model
c7 = LinearRegression()

# Initialize GridSearchCV with the model and parameters
grid = GridSearchCV(c7, parameters, cv=3)

# Fit the model to the data
grid.fit(x, y)

# Print the best parameters and the best estimator
print("Best Parameters:", grid.best_params_)
print("Best Estimator:", grid.best_estimator_)
print("Best Score:", grid.best_score_)

# Append the best estimator to the models list
models.append(grid.best_estimator_)

# Append the best score to the scores list
scores.append(grid.best_score_)


Best Parameters: {'fit_intercept': True}
Best Estimator: LinearRegression()
Best Score: 0.511615927217204


In [33]:
# Create a dataframe to store results
data=pd.DataFrame({"model_name":models,"best_score":scores})

In [34]:
print(data)

           model_name  best_score
0  LinearRegression()    0.511616


The results of the GridSearchCV indicate the following:

Best Parameters= ('fit_intercept': True),
Best Estimator= LinearRegression(),
Best Score= 0.7031337342269555
This suggests that using LinearRegression with fit_intercept=True resulted in the highest cross-validated score, which is approximately 0.7031

### Advance Track Capstone Project Task 3 and 4

In [35]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import metrics as m
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV

In [36]:
df1=pd.read_csv('cancer-data.csv')
df1 

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0            1        17.99         10.38          122.80     1001.0   
1            1        20.57         17.77          132.90     1326.0   
2            1        19.69         21.25          130.00     1203.0   
3            1        11.42         20.38           77.58      386.1   
4            1        20.29         14.34          135.10     1297.0   
..         ...          ...           ...             ...        ...   
564          1        21.56         22.39          142.00     1479.0   
565          1        20.13         28.25          131.20     1261.0   
566          1        16.60         28.08          108.30      858.1   
567          1        20.60         29.33          140.10     1265.0   
568          0         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0           0.2419  ...        25.380          17.33           184.60   
1           0.1812  ...        24.990          23.41           158.80   
2           0.2069  ...        23.570          25.53           152.50   
3           0.2597  ...        14.910          26.50            98.87   
4           0.1809  ...        22.540          16.67           152.20   
..             ...  ...           ...            ...              ...   
564         0.1726  ...        25.450          26.40           166.10   
565         0.1752  ...        23.690          38.25           155.00   
566         0.1590  ...        18.980          34.12           126.70   
567         0.2397  ...        25.740          39.42           184.60   
568         0.1587  ...         9.456          30.37            59.16   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0        2019.0           0.16220            0.66560           0.7119   
1        1956.0           0.12380            0.18660           0.2416   
2        1709.0           0.14440            0.42450           0.4504   
3         567.7           0.20980            0.86630           0.6869   
4        1575.0           0.13740            0.20500           0.4000   
..          ...               ...                ...              ...   
564      2027.0           0.14100            0.21130           0.4107   
565      1731.0           0.11660            0.19220           0.3215   
566      1124.0           0.11390            0.30940           0.3403   
567      1821.0           0.16500            0.86810           0.9387   
568       268.6           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    .

In [37]:
df1.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0          1        17.99         10.38          122.80     1001.0   
1          1        20.57         17.77          132.90     1326.0   
2          1        19.69         21.25          130.00     1203.0   
3          1        11.42         20.38           77.58      386.1   
4          1        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 31 columns]

In [38]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   diagnosis                569 non-null    int64  
 1   radius_mean              569 non-null    float64
 2   texture_mean             569 non-null    float64
 3   perimeter_mean           569 non-null    float64
 4   area_mean                569 non-null    float64
 5   smoothness_mean          569 non-null    float64
 6   compactness_mean         569 non-null    float64
 7   concavity_mean           569 non-null    float64
 8   concave points_mean      569 non-null    float64
 9   symmetry_mean            569 non-null    float64
 10  fractal_dimension_mean   569 non-null    float64
 11  radius_se                569 non-null    float64
 12  texture_se               569 non-null    float64
 13  perimeter_se             569 non-null    float64
 14  area_se                  5

In [39]:
df1.describe()

diagnosis  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  569.000000   569.000000    569.000000      569.000000   569.000000   
mean     0.372583    14.127292     19.289649       91.969033   654.889104   
std      0.483918     3.524049      4.301036       24.298981   351.914129   
min      0.000000     6.981000      9.710000       43.790000   143.500000   
25%      0.000000    11.700000     16.170000       75.170000   420.300000   
50%      0.000000    13.370000     18.840000       86.240000   551.100000   
75%      1.000000    15.780000     21.800000      104.100000   782.700000   
max      1.000000    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

The data indicates that there was a total of 569 diagnosis. Of these, 0 was the minimum recorded of concavity mean, concave point mean and concave worst and worst points. Concavity_mean can be 0 in a cancer dataset, indicating that there are no concave portions observed in the contour of the cell nuclei being measured. Concave worst represents the most severe case recorded in the dataset for this feature. This metric is used to help assess the aggressiveness of cancer cells and can be 0.

In [40]:
# data cleaning
df1.isna().sum()

diagnosis                  0
radius_mean                0
texture_mean               0
perimeter_mean             0
area_mean                  0
smoothness_mean            0
compactness_mean           0
concavity_mean             0
concave points_mean        0
symmetry_mean              0
fractal_dimension_mean     0
radius_se                  0
texture_se                 0
perimeter_se               0
area_se                    0
smoothness_se              0
compactness_se             0
concavity_se               0
concave points_se          0
symmetry_se                0
fractal_dimension_se       0
radius_worst               0
texture_worst              0
perimeter_worst            0
area_worst                 0
smoothness_worst           0
compactness_worst          0
concavity_worst            0
concave points_worst       0
symmetry_worst             0
fractal_dimension_worst    0
dtype: int64

In [41]:
df1.corr()

diagnosis  radius_mean  texture_mean  perimeter_mean  \
diagnosis                 1.000000     0.730029      0.415185        0.742636   
radius_mean               0.730029     1.000000      0.323782        0.997855   
texture_mean              0.415185     0.323782      1.000000        0.329533   
perimeter_mean            0.742636     0.997855      0.329533        1.000000   
area_mean                 0.708984     0.987357      0.321086        0.986507   
smoothness_mean           0.358560     0.170581     -0.023389        0.207278   
compactness_mean          0.596534     0.506124      0.236702        0.556936   
concavity_mean            0.696360     0.676764      0.302418        0.716136   
concave points_mean       0.776614     0.822529      0.293464        0.850977   
symmetry_mean             0.330499     0.147741      0.071401        0.183027   
fractal_dimension_mean   -0.012838    -0.311631     -0.076437       -0.261477   
radius_se                 0.567134     0.679090      0.275869        0.691765   
texture_se               -0.008303    -0.097317      0.386358       -0.086761   
perimeter_se              0.556141     0.674172      0.281673        0.693135   
area_se                   0.548236     0.735864      0.259845        0.744983   
smoothness_se            -0.067016    -0.222600      0.006614       -0.202694   
compactness_se            0.292999     0.206000      0.191975        0.250744   
concavity_se              0.253730     0.194204      0.143293        0.228082   
concave points_se         0.408042     0.376169      0.163851        0.407217   
symmetry_se              -0.006522    -0.104321      0.009127       -0.081629   
fractal_dimension_se      0.077972    -0.042641      0.054458       -0.005523   
radius_worst              0.776454     0.969539      0.352573        0.969476   
texture_worst             0.456903     0.297008      0.912045        0.303038   
perimeter_worst           0.782914     0.965137      0.358040        0.970387   
area_worst                0.733825     0.941082      0.343546        0.941550   
smoothness_worst          0.421465     0.119616      0.077503        0.150549   
compactness_worst         0.590998     0.413463      0.277830        0.455774   
concavity_worst           0.659610     0.526911      0.301025        0.563879   
concave points_worst      0.793566     0.744214      0.295316        0.771241   
symmetry_worst            0.416294     0.163953      0.105008        0.189115   
fractal_dimension_worst   0.323872     0.007066      0.119205        0.051019   

                         area_mean  smoothness_mean  compactness_mean  \
diagnosis                 0.708984         0.358560          0.596534   
radius_mean               0.987357         0.170581          0.506124   
texture_mean              0.321086        -0.023389          0.236702   
perimeter_mean            0.986507         0.207278          0.556936   
area_mean                 1.000000         0.177028          0.498502   
smoothness_mean           0.177028         1.000000          0.659123   
compactness_mean          0.498502         0.659123          1.000000   
concavity_mean            0.685983         0.521984          0.883121   
concave points_mean       0.823269         0.553695          0.831135   
symmetry_mean             0.151293         0.557775          0.602641   
fractal_dimension_mean   -0.283110         0.584792          0.565369   
radius_se                 0.732562         0.301467          0.497473   
texture_se               -0.066280         0.068406          0.046205   
perimeter_se              0.726628         0.296092          0.548905   
area_se                   0.800086         0.246552          0.455653   
smoothness_se            -0.166777         0.332375          0.135299   
compactness_se            0.212583         0.318943          0.738722   
concavity_se              0.207660         0.248396          0.570517   
concave points_se         0.372320         0.

In [42]:
#Extracting Independent and dependent Variable  

x=df1.iloc[:,1:31]
y=df1.iloc[:,0].values
#y=df1.diagnosis
print(x.shape)
print(y.shape)


(569, 30)
(569,)


In [43]:
# split the training and testing data
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)

In [44]:
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(455, 30)
(455,)
(114, 30)
(114,)


In [45]:
from sklearn.neighbors import KNeighborsClassifier #Import library
KNN = KNeighborsClassifier(n_neighbors = 15)

KNN.fit(x_train,y_train)                          #Train the Model
y_pred_KNN = KNN.predict(x_test)                  #Make Predictions
print(classification_report(y_test,y_pred_KNN))   #Evaluate the Model

              precision    recall  f1-score   support

           0       0.90      0.99      0.94        72
           1       0.97      0.81      0.88        42

    accuracy                           0.92       114
   macro avg       0.94      0.90      0.91       114
weighted avg       0.93      0.92      0.92       114



Class 0 (Precision: 0.90, Recall: 0.99, F1-Score: 0.94):
The model performs exceptionally well for class 0, with high recall and a strong F1-score.
Class 1 (Precision: 0.97, Recall: 0.81, F1-Score: 0.88):
The model also shows strong precision for class 1, although recall is lower than for class 0, resulting in a slightly lower F1-score.
The overall accuracy of 0.92 is strong, indicating that the model correctly classifies the majority of instances.

### Task4

In [46]:
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.model_selection import GridSearchCV

In [47]:
models=[]   # model[] is empty
scores=[]   #scores[] is empty
            #parameters = {'n_neighbors': [3, 5, 7, 9, 11]}   
            #grid = GridSearchCV(knn, parameters, cv=10
            #'metric':['manhattan','euclidean']} 

In [48]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

# Define the parameter grid with the values you want to test
param_grid = {'n_neighbors': [3, 5, 7, 9, 11],
                'metric': ['manhattan', 'euclidean']}

# Initialize the KNeighborsClassifier
knn = KNeighborsClassifier()

# Initialize GridSearchCV with the classifier, parameter grid, and cross-validation strategy
grid_search = GridSearchCV(knn, param_grid, cv=5)

# Fit the grid search to  training data
grid_search.fit(x_train, y_train)

# Get the best estimator found by the grid search
best_knn = grid_search.best_estimator_

# Use the best estimator to make predictions
y_pred_KNN = best_knn.predict(x_test)

# Print the classification report
print(classification_report(y_test, y_pred_KNN))

# Optionally, print the best parameters
print(f"Best parameters found: {grid_search.best_params_}")


              precision    recall  f1-score   support

           0       0.93      0.99      0.96        72
           1       0.97      0.88      0.93        42

    accuracy                           0.95       114
   macro avg       0.95      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114

Best parameters found: {'metric': 'manhattan', 'n_neighbors': 7}


These results suggest that the model performs well with high accuracy, precision, recall, and F1-scores, especially for class 0, while its performance on class 1 is still strong but slightly less effective in terms of recall. The best parameters indicate that a KNN model with 7 neighbors and Manhattan metric was found to be optimal for this classification task.
For class 0=
Precision: 0.93
93% of the instances predicted as class 0 were actually class 0.
Recall: 0.99
99% of the actual class 0 instances were correctly identified.
F1-Score: 0.96
The harmonic mean of precision and recall for class 0.
Support: 72
The number of actual instances of class 0.

<h3 style='color:red'> # Comments </h3>
    <h4 style='color:red'> 1)Task 1 (20 marks)  </h4> 
<h5 style='color:blue'> 

<br>---Learner has split the data into train test data correctly.</br>
<br>---Learner has correctly built the linear regression model using the training data. </br>
<br>---Learner has tested the model on test data too.</br>
<br>---Performance of the model is checked using various evaluation metrics like MAE, MSE, RMSE, and R-Squared score.</br>
<br>---Learner has provided relevant comments at relevant places.</br>
</h5>
<h5 style='color:green'>                                   
    <u> Marks scored:20 </u><br/>
-----------------------------------------------------------------------------------------------------------------------------------------------------------------    
</h5>    
 
<h3 style='color:red'> # Comments </h3>
    <h4 style='color:red'> 2)Task 2 (10 marks)  </h4> 
<h5 style='color:blue'>                                                                                
<br>---Learner has provided the code to build model and applied GridSearchCV for hyperparameter tuning. Created dataframe to store and display the results. </br>
<br>---Learner has provided relevant comments at relevant places. </br>


                           
<h5 style='color:green'>                                   
<u> Marks scored:10 </u><br/>
-----------------------------------------------------------------------------------------------------------------------------------------------------------------    
</h5> 
</h5> 

<h3 style='color:red'> # Comments </h3>
    <h4 style='color:red'> 1)Task 3 (20 marks)  </h4> 
<h5 style='color:blue'> 
<br>---Learner has correctly identified the target variable and the feature sets. </br>
<br>---Learner has split the data into train test data correctly.</br>
<br>---Learner has rightly built the classification model using the training data. </br>
<br>---Learner has tested the model on test data too.</br>
<br>---Learner has also analysed the classification report. Suggestion: Analyse the confusion matrix too.</br>
<br>---Learner has provided relevant comments at relevant places.</br>
</h5>
<h5 style='color:green'>                                   
    <u> Marks scored:20 </u><br/>
-----------------------------------------------------------------------------------------------------------------------------------------------------------------    
</h5>    
 
<h3 style='color:red'> # Comments </h3>
    <h4 style='color:red'> 2)Task 4 (10 marks)  </h4> 
<h5 style='color:blue'>                                                                                
<br>---Learner has provided the code to build KNN classification model and applied GridSearchCV for hyperparameter tuning. Created dataframe to store and display the results. </br>
<br>---Learner has provided relevant comments at relevant places. </br>
                           
<h5 style='color:green'>                                   
<u> Marks scored:10 </u><br/>
-----------------------------------------------------------------------------------------------------------------------------------------------------------------    
</h5> 
</h5> 

    
<h3 style='color:red'> # Comments </h3>
    <h4 style='color:red'> Overall Comments and Total Marks:</h4>
    
<h5 style='color:blue'>                                                                                
---Well-done! Your attention to the details and understanding of the code is much appreciated. <br>
---Keep up the good work and keep practicing. Happy coding!  </br>
<h5 style='color:green'>                                   
<u> Total Marks scored:60 </u><br/>
------------------------------------------------------------------------------------------------------------------------------------------------------------------    
</h5> 